Fix shipping cost formatting to respect shipping formula

ilyasfoo opened the pull request:
### Changes proposed in this Pull Request:

Closes #42892:

- Fix number formatting function to respect shipping formula
- Add relevant unit tests

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. In a fresh store with default country and currency configuration, go to WooCommerce > Settings > Shipping
2. Create a new shipping zone
3. From the new shipping zone, add a shipping method, choose `Flat rate`
4. In the `Cost` field, set `10*[qty]+(1+(2/[cost]))+([fee percent="10" max_fee="8.33"])`
5. Click `Create`
6. Refresh the page
7. Click on `Edit` on the new shipping method
8. Observe that the value is `10*[qty]+(1+(2/[cost]))+([fee percent="10" max_fee="8.33"])`
9. Test all possible combination of valid shipping formula and ensure the formatting does not break the formula

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

ilyasfoo committed:
Fix shipping cost formatting to respect formula

ilyasfoo committed:
Changelog

ilyasfoo committed:
Update test

psealock commented on plugins/woocommerce-admin/client/wp-admin-scripts/shipping-settings-region-picker/currency-context.js:
```diff
@@ -5,14 +5,39 @@ import { useContext, useEffect } from '@wordpress/element';
+		 * g: Global flag to match all instances in the string.
+		 */
+		return number.replace( /(\b\d+(\.\d+)?\b)(?![^[]*\])/g, ( n ) =>
+			numberFormat( config, n )
```
I'm not entirely sure we should format numbers in the formula.

`50000000.00 + (([qty]*2.00+1.00)(5.00*10.00))(1.00)` becomes the following. Its probably rare but I'm not sure how much we gain here. 

```
50,000,000.00 + (([qty]*2.00+1.00)(5.00*10.00))(1.00)
```

psealock replied:
Then again, it does seem to work with the formatted values. I'm still of the mind that this introduces complexity and we can remove it. What do you think?

<img width="502" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/5aad43d9-485a-4ad5-b2cd-a4440f854e79">

ilyasfoo replied:
Good point, I tend to agree now. I'll simplify to only handle formatting when it's exclusively a number. Thanks for the feedback!

psealock reviewed (commented):
Nice solution here @ilyasfoo 

I've tested a few combinations and seen them correctly reflected on the frontend at checkout.

I did raise an issue with formatting numbers in the formula.

psealock commented:
@ilyasfoo Something else is amiss here.

When setting my currency to Euro with `.` thousandth separator and `,` as decimal it treats it like a string

<img width="693" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/2015d506-4394-471e-bab2-cf35eb7b6cde">

`7,5` -> `7,00,5,00`
`7.5` -> `75,00`

and if you edit/save without changing the value changes each time.

ilyasfoo committed:
Reduce complexity by only formatting numbers, add support to handle arbitrary thousands and decimal separators

ilyasfoo commented:
Great catch, @psealock!

In https://github.com/woocommerce/woocommerce/pull/42916/commits/1c1da9bd0986ff8c9702eca0dacd98c768ed65da, I've converted the regex to be dynamic and respects the thousands and decimal separator set in config.

I also addressed your feedback on formatting formula. Now, the formatter only applies format when it's a single number. Please take another look! 🙇

ilyasfoo committed:
Update tests

ilyasfoo committed:
Fix regex to be more strict on numbers to match currency config

ilyasfoo committed:
Revert change that prevents 7.5 with thousands separator set to . to persist as 7.5 instead of 75

ilyasfoo committed:
Update tests

ilyasfoo commented on plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php:
```diff
@@ -281,7 +281,7 @@ public function find_shipping_classes( $package ) {
 		$value = is_null( $value ) ? '' : $value;
 		$value = wp_kses_post( trim( wp_unslash( $value ) ) );
-		$value = str_replace( array( get_woocommerce_currency_symbol(), html_entity_decode( get_woocommerce_currency_symbol() ), wc_get_price_thousand_separator() ), '', $value );
+		$value = str_replace( array( get_woocommerce_currency_symbol(), html_entity_decode( get_woocommerce_currency_symbol() ) ), '', $value );
```
This was interfering with saving an "invalid" value, in essence trying to outsmart the user:

1. Set thousands separator to `.`
2. Set decimal separator to `,`
3. Attempt to save `7.5`
4. Now it should still save `7.5` instead of `75`

p1702984979539049/1702957440.967159-slack-C01SFMVEYAK

psealock reviewed (approved):
This is testing well, thanks for all the hard work getting this right @ilyasfoo

psealock committed:
Handle different separators in free shipping and local pickup

psealock commented:
[This change](https://github.com/woocommerce/woocommerce/pull/42916#pullrequestreview-1789057553) had me thinking its not being done right in free shipping and local pickup either. One reason is `$value` is being changed before saving, but what really needs to happen is the decimal separator swapped out for `.` and the thousandth separator removed only for testing purposes but not for saving the value to the database.

In https://github.com/woocommerce/woocommerce/pull/42916/commits/b6a08cfd212368dc509a61acda7dec697c25e254 I fixed up the validation for free shipping and local pickup.

### Testing instructions addendum

1. Change the settings to different thousandth separator and decimal separator and currency.
2. Add free shipping with a requirement so the cost field shows.
3. Add a local pickup with cost too,
4. Add a flat rate and try using a formula
5. Save all, try entering bad numbers, but ultimately save values with different thousandth separator and decimal separator present in the string.
6. Add a product and go to checkout to make sure all the values are correctly reflected.

<img width="498" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/e3eeafc7-7960-4a28-938c-6d9b6546b9a5">

psealock committed:
fix linter errors

chihsuan reviewed (approved):
Thanks for the fix! @ilyasfoo @psealock 👍 

LGTM and confirmed the formatting does not break the formula.
